tools/fanatec_led_server.py

```diff
@@ -122,38 +141,51 @@ def run(self):
             if not self.tick():
                 break
 
-            if self.revLightsPercent > 0.95:
+            if self._revLightsPercent > 0.95:
                 rpms_maxed += 1
             else:
                 rpms_maxed = 0
 
             if rpms_maxed % 2 == 1:
-                leds = [False]*9
+                leds = [False] * 9
             else:
-                leds = [i/LEDS < self.revLightsPercent for i in range(LEDS)]
+                leds = [i / LEDS < self.revLightsPercent for i in range(LEDS)]
 
             if self.dbus:
                 set_leds(leds)
-                set_display(eval('self.%s'%self.display))
+                set_display(eval("self.%s" % self.display))
                 set_pedals_rumble(self.absInAction, self.tcInAction)
             else:
-                rumble = '%i'%( (0xff if self.tcInAction else 0) << 16 | (0xff if self.absInAction else 0) << 8 )
+                rumble = "%i" % (
+                    (0xFF if self.tcInAction else 0) << 16
+                    | (0xFF if self.absInAction else 0) << 8
+                )
 
                 if display is not None:
                     display_val = str(self.speedKmh)
-                    if self.display == 'gear':
-                        gear = {-1: "R", 0: "N"}
-                        display_val = gear[self.gear] if self.gear in gear else str(self.gear)
-                    open(display, 'w').write(display_val)
-
-                if pedals is not None: open(pedals, 'w').write(str(rumble))
-                fanatec_input.CSLEliteWheel.set_sysfs_rpm(leds, self.device)
+                    if self.display == "gear":
+                        if self.wheel == fanatec_input.CSLEliteWheel:
+                            gear = {-1: "R", 0: "N"}
+                        elif self.wheel == fanatec_input.CSLP1V2Wheel:
+                            gear = {-1: "111", 0: "000", 15: "12"}
```

Comment:
I don't understand these numbers.. Can you elaborate? Is this ams2 specific? The `gear` value here should be set in similar fashion by all Clients.

Reply:
So the problem (I mention this in another comment) is that different wheels have different LED behaviour. Presumable your wheel could be connected to a CSL DD base, or to  e.g. a Podium DD1. The P1V2Wheel, (and I think the basic BMW wheel) have just an LED display with three seven-segment displays, rather than the 9LEDS on your CSW base. you can write string characters to these but it appears that it only writes them if they are numbers. There will be some other approach to the Mclaren GT3 Wheel, and I'd need someone who has one to let me know exactly what given updates looks like on those screens. Hence, you don't get anything meaningful if you're in reverse. as a result I pushed these in. In AMS2, reverse gets 15 (FF) as its "gear byte". 12 looks most like an R on the wheel, but we could leave it as something else.

This is also why I differentiate between wheels rather than bases, as you mention in other comments... please let me know what you think about these choices.  The different wheels I think have more of an effect on the LED behaviour than the base.

Let me know what you think.

Reply:
I think to make this really correct, the driver has to expose the 9LEDs separate for the wheel and the wheelbase, the protocol would allow this. But I would postpone that and ignore the LEDs on the wheelbases for now, and just focus on LEDs/display on the wheel. So I totally agree with the approach.
But this is also why I would like to refactor `CSL_DD_P1_V2_WHEEL_ID` to  `CSL_DD_WHEELBASE_ID`, since this ID only refers to the wheelbase and does not give information on the mounted wheel.

Then, `self.wheel` should be set in the constructor of `class Client` by looking up the `wheel_id` sysfs file and creating the matching wheel object (s.t. like `self.wheel = {CSL_STEERING_WHEEL_P1_V2: fanatec_input.CSLP1V2Wheel, CLUBSPORT_STEERING_WHEEL_F1_IS_ID: fanatec_input.CSLEliteWheel}[wheel_id]`). See https://github.com/gotzl/hid-fanatecff/blob/next/hid-ftec.h for known wheel ids, but just the two you have atm with one of them as default should suffice for now. So, not really much different from what you have already, but just automatically creating the correct wheel object based on the mounted wheel ...

Then I think every game-client should present the same number scheme for the gears at this point, and I think  `-1 -> R` and `0 -> N` makes sense... So AMSClient should already map the `15` to `-1` internally, if this is what it sends for reverse. This makes me wonder, what does `-1` stand for in your code?

I hope you can follow me, don't hesitate to ask or propose s.t. different!

Reply:
Oh gosh, I just looked at the file you linked and it all makes sense now - I see that my 0020 was my wheelbase but i had conflated it with the wheel. I hope it is alright if I add all those as available constants just to keep clarity and allow for extensibility in the future.

Reply:
> which is I guess a nice way to represent reverse perhaps?

Well, my F1 wheel is able to represent 'reverse' with 'R' :) 
(Also, the behavior `-1 turns the lights off` changed in newest version of the driver, I think .. No, writing an empty string should turn off the lights.. I forgot to update the readme.)

> I am not used to having others read my code, so I've been a bit sloppy I think and left lots of things in I should have tidied up. I apologise for this - thank you for being patient with it.

No need to apologize!

> I hope it is alright if I add all those as available constants just to keep clarity and allow for extensibility in the future.

Sure, go ahead!